### DISCUSSION TITLE
cc2652P (Z-Stack_3.x.0) module support

### DISCUSSION THREAD

**Discussion by xhoobin** [arendst/Tasmota#10714] [Ideas]
### DESCRIPTION

cc2652P flashed into Z-Stack_3.x.0 firmware (https://github.com/Koenkk/Z-Stack-firmware/blob/master/coordinator/Z-Stack_3.x.0/bin/CC1352P_2_20201026.zip)
This firmware's major = 0x02 minor=0x07

  Unfortunately, tasmota does not support this version of the Z-Stack serial protocol, and cannot use the latest version of firmware and devices with higher performance.
Hope to support

### SCREENSHOTS
![image](https://user-images.githubusercontent.com/13211717/106000036-57b41c00-60e9-11eb-90c9-b5c73cc219a2.png)

**Comment by Jason2866** [comment]
Tasmota supports no cc2652 in general. For zigbee 3.0 the zbbridge can be used.

**Reply by s-hadinger** [reply]
There is no plan to support Z_Stack 3. Most of the effort is done for Ember ZNet (EZSP) for ZBBridge. We are favoring commonly available hardware like ZBBridge.

Edit: the main difference probably lies in the correct initialization sequence. If you want to invest time on it you are more than welcome to provide a PR. I can help you to understand the initialization state machine.

**Reply by xhoobin** [reply]
在提出此新功能支持前，我已经花少了时间来尝试支持Z_Stack3，但失败了。 主要是目前对Z_Stack3的串口协议改动和Tasmota本身的代码还不熟悉。 如果有人愿意协助的话， 我乐意投入时间去支持Z_Stack3

Before proposing this new feature support, I have spent less time trying to support Z_Stack3, but failed. The main reason is that the serial protocol changes of Z_Stack3 and Tasmota's own code are not yet familiar. If anyone is willing to assist, I am happy to invest time to support Z_Stack3

**Comment by xhoobin** [comment]
cc2652P（Z-Stack_3.x.0）在zigbee2mqtt中工作良好。我准备尝试学习zigbee2mqtt的代码来看如何在Tasmota中支持Z_Stack3。
我选择Tasmota的原因是Tasmota不像zigbee2mqtt一样，Tasmota可以在WIFI模块（ ESP8266，ESP32）中完成Z_Stack到MQTT的转换

cc2652P (Z-Stack_3.x.0) works well in zigbee2mqtt. I am going to try to learn the code of zigbee2mqtt to see how to support Z_Stack3 in Tasmota.
The reason I choose Tasmota is that Tasmota is not like zigbee2mqtt, Tasmota can complete the conversion from Z_Stack to MQTT in the WIFI module (ESP8266, ESP32)

**Reply by s-hadinger** [reply]
I can definitely help. But I'm not aware of a change in the Serial protocol. Can you remove the version test and see how the initialization is doing? I recall that Zigbee2MQTT has additional code at init for ZNP3 vs ZNP12

**Comment by s-hadinger** [comment]
Good news, it looks like ZStack 3 is mostly backwards compatible with ZNP1.2. I made a simple change to make ZStack 3 start. I didn't do any other test of pairing whatsoever.

#10732 10732

**Comment by s-hadinger** [comment]
I confirm with preliminary tests that Z-Stack 3.0 on CC2530 seems to be working fine: pairing worked with Aqara Temperature sensor. I will mark it as experimental in the doc.

@xhoobin Can you please confirm it is also working with cc2652P, I can't test on my side.

**Reply by hallard** [reply]
@s-hadinger not sure to have it all, would this means support of Z-Stack 3 makes use of CC2538 working also?

Because just wired one, joined aquara double switch and looks like it works fine :-)

```
00:00:02.754 WIF: Connected
00:00:03.007 HTP: Web server active on tasmota_040BC6-3014 with IP address 192.168.1.153
01:32:22.044 MQT: Attempting connection...
01:32:22.053 MQT: Connected
01:32:22.055 MQT: tele/tasmota_040BC6/LWT = Online (retained)
01:32:22.057 MQT: cmnd/tasmota_040BC6/POWER = 
01:32:22.063 MQT: tele/tasmota_040BC6/INFO1 = {"Module":"Generic","Version":"9.2.0.7(tasmota)","FallbackTopic":"cmnd/DVES_040BC6_fb/","GroupTopic":"cmnd/tasmotas/"}
01:32:22.070 MQT: tele/tasmota_040BC6/INFO2 = {"WebServerMode":"Admin","Hostname":"tasmota_040BC6-3014","IPAddress":"192.168.1.153"}
01:32:22.082 MQT: tele/tasmota_040BC6/INFO3 = {"RestartReason":"Software/System restart"}
01:32:24.483 QPC: Reset
01:32:26.395 MQT: tele/tasmota_040BC6/STATE = {"Time":"2021-02-15T01:32:26","Uptime":"0T00:00:10","UptimeSec":10,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Wifi":{"AP":1,"SSId":"CH2I-HOTSPOT","BSSId":"BC:4A:56:41:8D:90","Channel":1,"RSSI":54,"Signal":-73,"LinkCount":1,"Downtime":"0T00:00:04"}}
01:32:33.022 ZIG: rebooting CC2530 device
01:32:33.419 MQT: tele/tasmota_040BC6/RESULT = {"ZbState":{"Status":1,"Message":"CC2530 booted","RestartReason":"Power-up","MajorRel":2,"MinorRel":7}}
01:32:33.615 MQT: tele/tasmota_040BC6/RESULT = {"ZbState":{"Status":50,"MajorRel":2,"MinorRel":7,"MaintRel":2,"Revision":20200211}}
01:32:33.968 MQT: tele/tasmota_040BC6/RESULT = {"ZbState":{"Status":3,"Message":"Configured, starting coordinator"}}
01:32:34.269 MQT: tele/tasmota_040BC6/RESULT = {"ZbState":{"Status":40,"NewState":9,"Message":"Started as coordinator"}}
01:32:34.366 MQT: tele/tasmota_040BC6/RESULT = {"ZbState":{"Status":51,"IEEEAddr":"0x00124B00194AF396","ShortAddr":"0x0000","DeviceType":7,"DeviceState":9,"NumAssocDevices":0}}
01:32:34.916 MQT: tele/tasmota_040BC6/RESULT = {"ZbState":{"Status":0,"Message":"Started"}}
01:32:34.917 ZIG: Zigbee started
01:32:34.918 ZIG: No Zigbee device information in Flash
01:44:25.055 CMD: ZbPermitJoin 1
01:44:25.062 MQT: stat/tasmota_040BC6/RESULT = {"ZbPermitJoin":"Done"}
01:45:23.960 MQT: tele/tasmota_040BC6/RESULT = {"ZbState":{"Status":34,"IEEEAddr":"0x00158D000237BB46","ShortAddr":"0x4DDE","ParentNetwork":"0x0000"}}
01:45:24.512 MQT: tele/tasmota_040BC6/RESULT = {"ZbState":{"Status":30,"IEEEAddr":"0x00158D000237BB46","ShortAddr":"0x4DDE","PowerSource":false,"ReceiveWhenIdle":false,"Security":false}}
01:45:24.567 MQT: tele/tasmota_040BC6/SENSOR = {"ZbReceived":{"0x4DDE":{"Device":"0x4DDE","ModelId":"lumi.sensor_86sw2","AppVersion":5,"Endpoint":1,"LinkQuality":120}}}
01:45:25.419 MQT: tele/tasmota_040BC6/SENSOR = {"ZbReceived":{"0x4DDE":{"Device":"0x4DDE","BatteryVoltage":3.05,"BatteryPercentage":100,"Endpoint":1,"LinkQuality":118}}}
01:45:26.662 ZIG: Zigbee Devices Data store in Flash (0x402FF800 - 33 bytes)
01:45:39.552 MQT: tele/tasmota_040BC6/SENSOR = {"ZbReceived":{"0x4DDE":{"Device":"0x4DDE","Power":1,"Power2":0,"Endpoint":1,"LinkQuality":128}}}
01:45:44.645 MQT: tele/tasmota_040BC6/SENSOR = {"ZbReceived":{"0x4DDE":{"Device":"0x4DDE","Power":1,"Power2":0,"Endpoint":1,"LinkQuality":126}}}
01:45:50.699 MQT: tele/tasmota_040BC6/SENSOR = {"ZbReceived":{"0x4DDE":{"Device":"0x4DDE","Power":1,"Power2":0,"Endpoint":2,"LinkQuality":120}}}
01:45:56.905 MQT: tele/tasmota_040BC6/SENSOR = {"ZbReceived":{"0x4DDE":{"Device":"0x4DDE","Power":1,"Power2":0,"Endpoint":3,"LinkQuality":136}}}
01:46:06.412 MQT: tele/tasmota_040BC6/SENSOR = {"ZbReceived":{"0x4DDE":{"Device":"0x4DDE","Power":1,"Power2":0,"Endpoint":3,"LinkQuality":144}}}
01:46:15.874 MQT: tele/tasmota_040BC6/SENSOR = {"ZbReceived":{"0x4DDE":{"Device":"0x4DDE","Power":1,"Power2":0,"Endpoint":1,"LinkQuality":134}}}
01:46:30.345 MQT: tele/tasmota_040BC6/SENSOR = {"ZbReceived":{"0x4DDE":{"Device":"0x4DDE","Power":1,"Power2":0,"Endpoint":2,"LinkQuality":113}}}
```

**Reply by s-hadinger** [reply]
Well. It look it does. I don't have a cc2538 and I'm glad to see it works. 

Can you please update the documentation?

**Reply by hallard** [reply]
Sure, I created a github project for the PCB [here](https://github.com/hallard/cc2538-multi-connect) I will now update tasmota doc according to.

**Reply by s-hadinger** [reply]
Great. Do you have any plans for esp32 equivalent? This would bring a lot of possibilities

**Reply by hallard** [reply]
you can plug esp32 mini d1, same pinout

![1B67505E-25FC-4CB9-A3A4-99E045AB05AF](https://user-images.githubusercontent.com/2471931/108043299-3f2c8700-7041-11eb-9f47-1ec8cca5b601.jpeg)

**Reply by s-hadinger** [reply]
Wow, I didn't know it existed

**Reply by s-hadinger** [reply]
Last question, the ESP32 version seems larger. Will it fit?

**Reply by hallard** [reply]
Correct will not fit on enclosure, I need to fix that, honestly wasn't aware you needed it for ZigBee. But in all case you'll be able test before I review the board.
May I ask what features you're thinking about?

**Reply by s-hadinger** [reply]
Esp32 allows to make lots of automation directly on the chip without needing a separate gateway. And we now get very powerful scripting with Berry

**Reply by hallard** [reply]
Makes senses, not hyper confortable having all HA relying on one box and separate features with powerfull tasmota devices and scripting is another way to automate I'm thinking about for long time.
Thanks.

**Reply by s-hadinger** [reply]
I ordered some of those ESP32 D1 mini, the power supply is too weak and they do `Brownout detector was triggered`. I have doubts about their stability.

**Reply by Jason2866** [reply]
Found a circuit dagram for the ESP32 D1 mini. The Vreg is a ME6211. Max. output current 500mA.
This is not enough for a ESP32. https://www.google.com/url?sa=t&source=web&rct=j&url=https://www.espressif.com/sites/default/files/documentation/esp32_hardware_design_guidelines_en.pdf&ved=2ahUKEwj0w-L-pfHuAhWF2uAKHZnZCJMQFjAAegQIBBAC&usg=AOvVaw07R8p6nXT5PZWH2w74F0br&cshid=1613577951795

**Reply by hallard** [reply]
Well, 500mA is fair enough for a ESP32.
Adafruit ESP32 board have 500mA regulator and state ESP32 max is 250mA (same as Espressif datasheet), so may be you put some other things on 3V3 rail that have excessive consumption, but max power from CC2538+CC2592 is 190mA so even both working full power it's under 500mA (And I doubt all are max power in same time because it's just peak on transmit not full time)
But we can suspect of course Chineese regulator (and associated hardware) are not as good as the one used by Adafruit or Sparkfun because I also had sometime brownout trigger at startup, even with nothing connected, just ESP32

so sometime, I disable the brownout it in setup as follow :-)
```cpp
  // disable brownout detector reset
  WRITE_PERI_REG(RTC_CNTL_BROWN_OUT_REG, 0); 
```

PS : I had to disable brownout also with espressif dev kit but it makes sense, I had connected RGB Led matrix on 5V rail also and these are consuming a lot. Worth trying to add better (or higher) caps on 5V and 3V3 rail on these boards
